crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs

```diff
@@ -0,0 +1,147 @@
+use hir::Semantics;
+use ide_db::assists::{AssistId, AssistKind};
+use ide_db::base_db::FileId;
+use ide_db::RootDatabase;
+use syntax::ast::{self, HasName};
+use syntax::{AstNode, SyntaxKind};
+
+use crate::assist_context::{AssistContext, Assists};
+
+// Assist: convert_nested_function_to_closure
+//
+// Converts a function that is defined within the body of another function into a closure.
+//
+// ```
+// fn main() {
+//     fn foo(label: &str, $0number: u64) {
+//         println!("{}: {}", label, number);
+//     }
+//
+//     foo("Bar", 100);
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     let foo = |label: &str, number: u64| {
+//         println!("{}: {}", label, number);
+//     };
+//
+//     foo("Bar", 100);
+// }
+// ```
+pub(crate) fn convert_nested_function_to_closure(
+    acc: &mut Assists,
+    ctx: &AssistContext<'_>,
+) -> Option<()> {
+    let function = ctx.find_node_at_offset::<ast::Fn>()?;
+    if !is_nested_function(&function) {
+        return None;
+    }
+
+    let target = function.syntax().text_range();
+    let body = function.body()?;
+    let name = function.name()?;
+    let params = function.param_list()?;
+    let has_semicolon = has_semicolon(&function, ctx.file_id(), &ctx.sema);
```

Comment:
This check can be don in the `|edit|` closure, that way we only do the work if the assist gets executed